Fix double handling in `collect_tokens`

nnethercote opened the pull request:
Double handling of AST nodes can occur in `collect_tokens`. This is when an inner call to `collect_tokens` produces an AST node, and then an outer call to `collect_tokens` produces the same AST node. This can happen in a few places, e.g. expression statements where the statement delegates `HasTokens` and `HasAttrs` to the expression. It will also happen more after #124141.

This PR fixes some double handling cases that cause problems, including #129166.

r? @petrochenkov

nnethercote committed:
Avoid unnecessary `cloned`.

nnethercote committed:
Avoid nested replacement ranges.

In a case like this:
```
mod a {
    mod b {
        #[cfg_attr(unix, inline)]
        fn f() {
            #[cfg_attr(linux, inline)]
            fn g1() {}
            #[cfg_attr(linux, inline)]
            fn g2() {}
        }
    }
}
```
We currently end up with the following replacement ranges.
- The lazy tokens for `f` has replacement ranges for `g1` and `g2`.
- The lazy tokens for `a` has replacement ranges for `f`, `g1`, and
  `g2`.

I.e. the replacement ranges for `g1` and `g2` are duplicated. In
general, replacement ranges for inner AST nodes are duplicated up the
chain for each nested `collect_tokens` call. And the code that processes
the replacements is careful about the ordering in which the replacements
are applied, to ensure that inner replacements are applied before outer
replacements.

But all of this is unnecessary. If you apply an inner replacement and
then an outer replacement, the outer replacement completely overwrites
the inner replacement.

This commit avoids the duplication by removing replacements from
`self.capture_state.parser_replacements` when they are used. (The effect
on the example above is that the lazy tokesn for `a` no longer include
replacement ranges for `g1` and `g2`.) This eliminates the possibility
of nested replacements on individual AST nodes, which avoids the need
for careful ordering of replacements.

nnethercote committed:
Return earlier in some cases in `collect_token`.

This example triggers an assertion failure:
```
fn f() -> u32 {
    #[cfg_eval] #[cfg(not(FALSE))] 0
}
```
The sequence of events:
- `configure_annotatable` calls `parse_expr_force_collect`, which calls
  `collect_tokens`.
- Within that, we end up in `parse_expr_dot_or_call`, which again calls
  `collect_tokens`.
  - The return value of the `f` call is the expression `0`.
  - This inner call collects tokens for `0` (parser range 10..11) and
    creates a replacement covering `#[cfg(not(FALSE))] 0` (parser range
    0..11).
- We return to the outer `collect_tokens` call. The return value of the
  `f` call is *again* the expression `0`, again with the range 10..11,
  but the replacement from earlier covers the range 0..11. The code
  mistakenly assumes that any attributes from an inner `collect_tokens`
  call fit entirely within the body of the result of an outer
  `collect_tokens` call. So it adjusts the replacement parser range
  0..11 to a node range by subtracting 10, resulting in -10..1. This is
  an invalid range and triggers an assertion failure.

It's tricky to follow, but basically things get complicated when an AST
node is returned from an inner `collect_tokens` call and then returned
again from an outer `collect_token` node without being wrapped in any
kind of additional layer.

This commit changes `collect_tokens` to return early in some extra cases,
avoiding the construction of lazy tokens. In the example above, the
outer `collect_tokens` returns earlier because the `0` token already has
tokens and `self.capture_state.capturing` is `Capturing::No`. This early
return avoids the creation of the invalid range and the assertion
failure.

Fixes #129166. Note: these invalid ranges have been happening for a long
time. #128725 looks like it's at fault only because it introduced the
assertion that catches the invalid ranges.

nnethercote committed:
Avoid double-handling of attributes in `collect_tokens`.

By keeping track of attributes that have been previously processed.

This fixes the `macro-rules-derive-cfg.stdout` test, and is necessary
for #124141 which removes nonterminals.

Also shrink the `SmallVec` inline size used in `IntervalSet`. 2 gives
slightly better perf than 4 now that there's an `IntervalSet` in
`Parser`, which is cloned reasonably often.

nnethercote committed:
Tweak `tests/ui/proc-macro/macro-rules-derive-cfg.rs`.

- Trim some unnecessary fat from the type declaration.
- Add another attribute, to make it a stronger test of `cfg_attr`
  processing. Note that the current output is incorrect, because it
  duplicates the added attribute. The next commit will fix this.

nnethercote committed:
Clarify a comment.

nnethercote commented:
The last three commits are tricky and hard to review. Even after multiple weeks of working with this code, I still find it hard to understand and reason about, and I wish it could be made simpler. But the good news is that these commits are enough to the get final failing test (`tests/ui/proc-macros/macro-rules-derive-cfg.rs`) working in #124141.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10484872057/job/29039974156) [(plain)](https://github.com/rust-lang/rust/commit/a0f9bfd632bd24caabab1475df696b1cfc4cd822/checks/29039974156/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

running 232 tests
........................................................................................  88/232
........................................................................................ 176/232
........................2024-08-21T07:08:47.047358Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
[crashes] tests/crashes/129166.rs ... F
...............................

failures:
failures:

---- [crashes] tests/crashes/129166.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
thread '[crashes] tests/crashes/129166.rs' panicked at src/tools/compiletest/src/runtest.rs:368:18:
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
```

</details>

nnethercote commented:
I think this should get a crater run, just because I don't fully trust that the test suite is thoroughly testing `cfg_attr` stuff.

And why not do a perf run, just to check there's no regressions.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 7d6580971e8af6eb25b5fd670cfe0308795df041 with merge 8b22b49ea046aed5d405c5bda40972c039aa70b5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7d6580971e8af6eb25b5fd670cfe0308795df041","merge_sha":"8b22b49ea046aed5d405c5bda40972c039aa70b5"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10485687908/job/29042337358) [(plain)](https://github.com/rust-lang/rust/commit/7d6580971e8af6eb25b5fd670cfe0308795df041/checks/29042337358/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
1 error: removing an expression is not supported in this position
-   --> $DIR/invalid-node-range-issue-129166.rs:5:17
+   --> $DIR/invalid-node-range-issue-129166.rs:7:17
3    |
4 LL |     #[cfg_eval] #[cfg(not(FALSE))] 0


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/invalid-node-range-issue-129166/invalid-node-range-issue-129166.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/invalid-node-range-issue-129166/invalid-node-range-issue-129166.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args conditional-compilation/invalid-node-range-issue-129166.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/conditional-compilation/invalid-node-range-issue-129166.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/invalid-node-range-issue-129166" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/invalid-node-range-issue-129166/auxiliary"
--- stderr -------------------------------
error: removing an expression is not supported in this position
##[error]  --> /checkout/tests/ui/conditional-compilation/invalid-node-range-issue-129166.rs:7:17
   |
   |
LL |     #[cfg_eval] #[cfg(not(FALSE))] 0

error: aborting due to 1 previous error
------------------------------------------

```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10485720601/job/29047829277)
Build commit: 8b22b49ea046aed5d405c5bda40972c039aa70b5 (`8b22b49ea046aed5d405c5bda40972c039aa70b5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10485720601/job/29047829277"},"merge_sha":"8b22b49ea046aed5d405c5bda40972c039aa70b5"} -->

rust-timer commented:
Queued 8b22b49ea046aed5d405c5bda40972c039aa70b5 with parent 5aea14073eee9e403c3bb857490cd6aa4a395531, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5aea14073eee9e403c3bb857490cd6aa4a395531&end=8b22b49ea046aed5d405c5bda40972c039aa70b5).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (8b22b49ea046aed5d405c5bda40972c039aa70b5): [comparison URL](https://perf.rust-lang.org/compare.html?start=5aea14073eee9e403c3bb857490cd6aa4a395531&end=8b22b49ea046aed5d405c5bda40972c039aa70b5&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=5aea14073eee9e403c3bb857490cd6aa4a395531&end=8b22b49ea046aed5d405c5bda40972c039aa70b5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.7%]   | 43    |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.1%, 0.6%]   | 16    |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.3%, -0.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-0.3%, 0.7%]  | 44    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5aea14073eee9e403c3bb857490cd6aa4a395531&end=8b22b49ea046aed5d405c5bda40972c039aa70b5&stat=max-rss)
<details>
<summary>Results (secondary -2.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-3.4%, -2.0%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5aea14073eee9e403c3bb857490cd6aa4a395531&end=8b22b49ea046aed5d405c5bda40972c039aa70b5&stat=cycles:u)
<details>
<summary>Results (secondary -2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5aea14073eee9e403c3bb857490cd6aa4a395531&end=8b22b49ea046aed5d405c5bda40972c039aa70b5&stat=size:linked_artifact)
<details>
<summary>Results (secondary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

**Bootstrap**: 748.876s -> 749.555s (0.09%)
**Artifact size**: 338.91 MiB -> 338.99 MiB (0.02%)

petrochenkov commented on tests/ui/conditional-compilation/invalid-node-range-issue-129166.rs:
```diff
@@ -0,0 +1,11 @@
+// This was triggering an assertion failure in `NodeRange::new`.
```
Could you split the assert in `NodeRange::new` into two asserts?
Right now it is not clear which part of the condition actually fails.

petrochenkov commented:
I'll start crater once https://github.com/rust-lang/rust/pull/129346#discussion_r1725291124 is addressed and CI is green.
@rustbot author

nnethercote committed:
Split the assertion in `NodeRange::new`.

nnethercote committed:
Optimize `collect_tokens` a little.

Use `Cow` to avoid cloning `ret.attrs()` unless necessary. This requires
moving some things around to satisfy the borrow checker.

nnethercote commented:
I have adding a commit that avoids cloning `ret.attrs()` except when necessary. This gets back part of the perf regression. The rest is due to `Parser::clone` being more expensive because of the new `FxHashSet` in `CaptureState`. I tried changing it to a `GrowableBitSet` which seemed like it should have helped but didn't.

Let's see what the new status is:
@bors try @rust-time queue

bors commented:
:hourglass: Trying commit cab3178ee9c0503aaea9b6e2167375ea97dedbe7 with merge a70947b8062a88e6f76266f2540de6ea6c2bfba5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"cab3178ee9c0503aaea9b6e2167375ea97dedbe7","merge_sha":"a70947b8062a88e6f76266f2540de6ea6c2bfba5"} -->

lqd commented:
@rust-timer build a70947b8062a88e6f76266f2540de6ea6c2bfba5

rust-timer commented:
Queued a70947b8062a88e6f76266f2540de6ea6c2bfba5 with parent a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1&end=a70947b8062a88e6f76266f2540de6ea6c2bfba5).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.6 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (a70947b8062a88e6f76266f2540de6ea6c2bfba5): [comparison URL](https://perf.rust-lang.org/compare.html?start=a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1&end=a70947b8062a88e6f76266f2540de6ea6c2bfba5&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1&end=a70947b8062a88e6f76266f2540de6ea6c2bfba5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.2%, 0.7%] | 25    |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.2%, 0.6%] | 17    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.2%, 0.7%] | 25    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1&end=a70947b8062a88e6f76266f2540de6ea6c2bfba5&stat=max-rss)
<details>
<summary>Results (primary -0.8%, secondary 2.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.5%, 1.3%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [2.1%, 2.7%]   | 2     |
| Improvements ✅ <br /> (primary)   | -4.3% | [-4.3%, -4.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-4.3%, 1.3%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1&end=a70947b8062a88e6f76266f2540de6ea6c2bfba5&stat=cycles:u)
<details>
<summary>Results (secondary 2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.3%, 2.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a32d4a0e822a29a6682e08b75a8df4c29c7fa9f1&end=a70947b8062a88e6f76266f2540de6ea6c2bfba5&stat=size:linked_artifact)
<details>
<summary>Results (secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

**Bootstrap**: 750.761s -> 751.313s (0.07%)
**Artifact size**: 338.92 MiB -> 339.03 MiB (0.03%)

nnethercote commented:
That didn't help much, but I have another idea that I think should work better. I will try it tomorrow. We should hold off on a crater run until then.

nnethercote commented:
Another attempt to improve perf, this time using a `RangeSet` from the `rangemap` crate.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 0ec00d1c0be7fa4c3a3ae8036224f43d8c502b10 with merge 363237818fb36ba16d6989c67e9877d5c91f87b3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0ec00d1c0be7fa4c3a3ae8036224f43d8c502b10","merge_sha":"363237818fb36ba16d6989c67e9877d5c91f87b3"} -->

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10521183273/job/29151396107) [(plain)](https://github.com/rust-lang/rust/commit/0ec00d1c0be7fa4c3a3ae8036224f43d8c502b10/checks/29151396107/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.651 Building wheels for collected packages: reuse
#13 2.652   Building wheel for reuse (pyproject.toml): started
#13 2.897   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.898   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.898   Stored in directory: /tmp/pip-ephem-wheel-cache-c8l7_h4o/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.901 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.287 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.288 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.801 Collecting virtualenv
#13 3.801 Collecting virtualenv
#13 3.840   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 3.917      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 75.3 MB/s eta 0:00:00
#13 3.969 Collecting platformdirs<5,>=3.9.1
#13 3.972   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.004 Collecting filelock<4,>=3.12.2
#13 4.024 Collecting distlib<1,>=0.3.7
#13 4.027   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.034      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 94.2 MB/s eta 0:00:00
#13 4.034      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 94.2 MB/s eta 0:00:00
#13 4.116 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.303 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.4s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      208832 kB
DirectMap2M:     8179712 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/b5723af3457b9cd3795eeb97e9af2d34964854f2/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-b5723af3457b9cd3795eeb97e9af2d34964854f2-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
fmt: checked 5508 files
tidy check
Go to `src/tools/tidy/src/deps.rs:237` for the list.
tidy error: Dependency for . not explicitly permitted: registry+https://github.com/rust-lang/crates.io-index#rangemap@1.5.1
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 65.5 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Fri Aug 23 06:49:02 UTC 2024
  network time: Fri, 23 Aug 2024 06:49:02 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10521209986/job/29155771269)
Build commit: 363237818fb36ba16d6989c67e9877d5c91f87b3 (`363237818fb36ba16d6989c67e9877d5c91f87b3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10521209986/job/29155771269"},"merge_sha":"363237818fb36ba16d6989c67e9877d5c91f87b3"} -->

rust-timer commented:
Queued 363237818fb36ba16d6989c67e9877d5c91f87b3 with parent b5723af3457b9cd3795eeb97e9af2d34964854f2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b5723af3457b9cd3795eeb97e9af2d34964854f2&end=363237818fb36ba16d6989c67e9877d5c91f87b3).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (363237818fb36ba16d6989c67e9877d5c91f87b3): [comparison URL](https://perf.rust-lang.org/compare.html?start=b5723af3457b9cd3795eeb97e9af2d34964854f2&end=363237818fb36ba16d6989c67e9877d5c91f87b3&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=b5723af3457b9cd3795eeb97e9af2d34964854f2&end=363237818fb36ba16d6989c67e9877d5c91f87b3&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.4%]   | 13    |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | 0.2%  | [0.2%, 0.4%]   | 13    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b5723af3457b9cd3795eeb97e9af2d34964854f2&end=363237818fb36ba16d6989c67e9877d5c91f87b3&stat=max-rss)
<details>
<summary>Results (secondary 0.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [3.5%, 3.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b5723af3457b9cd3795eeb97e9af2d34964854f2&end=363237818fb36ba16d6989c67e9877d5c91f87b3&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=b5723af3457b9cd3795eeb97e9af2d34964854f2&end=363237818fb36ba16d6989c67e9877d5c91f87b3&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 749.494s -> 751.374s (0.25%)
**Artifact size**: 338.87 MiB -> 339.07 MiB (0.06%)

lqd commented:
> this time using a `RangeSet` from the `rangemap` crate.

Drive-by question, is that like `rustc_index`'s `IntervalSet`?

nnethercote commented:
> Drive-by question, is that like `rustc_index`'s `IntervalSet`?

Oh! It is, thank you for the suggestion. But `IntervalSet` is even better because its API matches that of `HashSet` exactly, at least for the uses in this PR. I have switched the code to use it, let's do another perf run:

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit fb6a47c0ddcf4d1794c86e1ad15016d857dda883 with merge 028775a58e49010f319d79f4712d82fb6899ac07...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fb6a47c0ddcf4d1794c86e1ad15016d857dda883","merge_sha":"028775a58e49010f319d79f4712d82fb6899ac07"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10526462384/job/29172836807)
Build commit: 028775a58e49010f319d79f4712d82fb6899ac07 (`028775a58e49010f319d79f4712d82fb6899ac07`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10526462384/job/29172836807"},"merge_sha":"028775a58e49010f319d79f4712d82fb6899ac07"} -->

rust-timer commented:
Queued 028775a58e49010f319d79f4712d82fb6899ac07 with parent c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a&end=028775a58e49010f319d79f4712d82fb6899ac07).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (028775a58e49010f319d79f4712d82fb6899ac07): [comparison URL](https://perf.rust-lang.org/compare.html?start=c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a&end=028775a58e49010f319d79f4712d82fb6899ac07&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a&end=028775a58e49010f319d79f4712d82fb6899ac07&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.3%, 0.5%] | 11    |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.2%, 1.4%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.3%, 0.5%] | 11    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a&end=028775a58e49010f319d79f4712d82fb6899ac07&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a&end=028775a58e49010f319d79f4712d82fb6899ac07&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a&end=028775a58e49010f319d79f4712d82fb6899ac07&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 751.695s -> 751.859s (0.02%)
**Artifact size**: 339.03 MiB -> 338.98 MiB (-0.01%)

nnethercote commented:
Ok, `IntervalSet` gives similar results to `RangeMap`, which is great! There is still a very small perf regression but it's less than it was, and the fix is necessary for correctness both now and even more when #124141 gets merged. So I think this is good enough for a crater run.

@bors try

bors commented:
:hourglass: Trying commit d4bf28c014772f7233fc396992164c72c08fbfd9 with merge 96ad39d809ea5a132d223acdaa01b9a8d3a26348...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d4bf28c014772f7233fc396992164c72c08fbfd9","merge_sha":"96ad39d809ea5a132d223acdaa01b9a8d3a26348"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10532309902/job/29189411454)
Build commit: 96ad39d809ea5a132d223acdaa01b9a8d3a26348 (`96ad39d809ea5a132d223acdaa01b9a8d3a26348`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10532309902/job/29189411454"},"merge_sha":"96ad39d809ea5a132d223acdaa01b9a8d3a26348"} -->

nnethercote commented:
@craterbot check

nnethercote commented:
The reason that `IntervalSet` works well is that multi-line `///` doc comments get split into one attribute per line. So it's common to have lots of attributes on an AST node, but those attributes all have consecutive `AttrId`s. Much nicer to have a single range like `50..100` instead of 50 individual entries.

I really wish that multiple `///` doc comment lines instead collapsed into a single attribute. This would make it easier to give good perf inside the compiler. I suspect it's not possible to change this because of proc macros.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit d4bf28c014772f7233fc396992164c72c08fbfd9 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov d4bf28c014772f7233fc396992164c72c08fbfd9 -->
<!-- homu: {"type":"Approved","sha":"d4bf28c014772f7233fc396992164c72c08fbfd9","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d4bf28c014772f7233fc396992164c72c08fbfd9 with merge 6d05f1217067474d0974ffed292a1a025d075826...
<!-- homu: {"type":"BuildStarted","head_sha":"d4bf28c014772f7233fc396992164c72c08fbfd9","merge_sha":"6d05f1217067474d0974ffed292a1a025d075826"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10757409454/job/29833275633)
Approved by: petrochenkov
Pushing 6d05f1217067474d0974ffed292a1a025d075826 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10757409454/job/29833275633"},"merge_sha":"6d05f1217067474d0974ffed292a1a025d075826"} -->

rust-timer commented:
Finished benchmarking commit (6d05f1217067474d0974ffed292a1a025d075826): [comparison URL](https://perf.rust-lang.org/compare.html?start=7f4b270aa49823ef142018e969fccb7eea4a7ea6&end=6d05f1217067474d0974ffed292a1a025d075826&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=7f4b270aa49823ef142018e969fccb7eea4a7ea6&end=6d05f1217067474d0974ffed292a1a025d075826&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.4%, -0.3%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7f4b270aa49823ef142018e969fccb7eea4a7ea6&end=6d05f1217067474d0974ffed292a1a025d075826&stat=max-rss)
<details>
<summary>Results (primary -1.9%, secondary 1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [2.6%, 2.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [2.3%, 4.2%]   | 7     |
| Improvements ✅ <br /> (primary)   | -3.4% | [-3.5%, -3.4%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.7%, -0.6%] | 3     |
| All ❌✅ (primary)                 | -1.9% | [-3.5%, 2.6%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7f4b270aa49823ef142018e969fccb7eea4a7ea6&end=6d05f1217067474d0974ffed292a1a025d075826&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=7f4b270aa49823ef142018e969fccb7eea4a7ea6&end=6d05f1217067474d0974ffed292a1a025d075826&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 754.658s -> 756.515s (0.25%)
**Artifact size**: 341.04 MiB -> 341.16 MiB (0.04%)

danielhenrymantilla commented:
FWIW, this did regress the following test: https://docs.rs/crate/cfg_eval/0.1.2/source/tests/item.rs

  - https://github.com/danielhenrymantilla/cfg_eval.rs/actions/workflows/future-proof.yml

I don't know why crater did not see it 😔

danielhenrymantilla commented on tests/ui/proc-macro/macro-rules-derive-cfg.rs:
```diff
@@ -14,17 +14,15 @@ extern crate test_macros;
 macro_rules! produce_it {
     ($expr:expr) => {
         #[derive(Print)]
-        struct Foo {
-            val: [bool; {
-                let a = #[cfg_attr(not(FALSE), rustc_dummy(first))] $expr;
-                0
-            }]
-        }
```
This case stopped working?

nnethercote replied:
Fixed in #132587.

Noratrieb commented:
even if crater found your crate, it was only a check run

lqd commented:
> even if crater found your crate, it was only a check run

and since beta runs are test runs, the regression is indeed present in the 1.83 unexpected test failures: https://crater-reports.s3.amazonaws.com/beta-1.83-1/beta-2024-10-19/gh/danielhenrymantilla.cfg_eval.rs/log.txt
